topi/python/topi/arm_cpu/depthwise_conv2d.py

```diff
@@ -5,15 +5,140 @@
+                       policy='candidate', candidate=[
+                           [n, co, oh, ow, kh, kw, vh, vw, vc],
+                           [n, co, oh, ow, kh, kw, vc, vh, vw]])
+
+    cfg.define_reorder("reorder_1",
+                       [n, co, oh, ow, vh, vw, vc],
+                       policy='candidate', candidate=[
+                           [n, co, oh, ow, vh, vw, vc],
+                           [n, co, oh, ow, vc, vh, vw],
+                           [n, co, oh, ow, vh, vc, vw]])
+
+    cfg.define_annotate("ann_reduce", [kh, kw], policy='try_unroll')
+    cfg.define_annotate("ann_spatial", [vh, vw, vc], policy='try_unroll_vec')
+    # ====================================================================
+
+    VC = cfg["tile_co"].size[-1]
+    VH = cfg["tile_oh"].size[-1]
+    VW = cfg["tile_ow"].size[-1]
+
+    dvshape = (N, OH // VH, OW // VW, C, VH*HSTR + KH-1, VW*WSTR + KW-1)
+    kvshape = (C // VC, M, KH, KW, VC)
+    ovshape = (N, C * M // VC, OH // VH, OW // VW, VH, VW, VC)
+    oshape = (N, C * M, OH, OW)
+
+    data_vec = tvm.compute(dvshape, lambda n, h, w, c, vh, vw:
+                           data_pad[n][c][h*VH*HSTR+vh][w*VW*WSTR+vw],
+                           name='data_vec')
+
+    if pre_packed:
+        kernel_vec = kernel
+    else:
+        kernel_vec = tvm.compute(kvshape, lambda co, m, kh, kw, vc:
+                                 kernel[co*VC+vc][m][kh][kw],
+                                 name='kernel_vec')
+
+    kh = tvm.reduce_axis((0, KH), name='kh')
+    kw = tvm.reduce_axis((0, KW), name='kw')
+
+    conv = tvm.compute(ovshape, lambda n, co, h, w, vh, vw, vc:
+                       tvm.sum(data_vec[n, h, w, (co * VC + vc) // M, vh * HSTR + kh,
+                                        vw * WSTR + kw].astype(out_dtype) *
+                               kernel_vec[co // M, co % M, kh, kw, vc].astype(out_dtype),
+                               axis=[kh, kw]), name='conv')
+
+    output = tvm.compute(oshape, lambda n, co, h, w:
+                         conv[n][co//VC][h//VH][w//VW][h%VH][w%VW][co%VC],
+                         name='output_unpack', tag='spatial_depthwise_conv_nchw_output',
+                         attrs={'workload': depthwise_conv_arg_to_workload(data, kernel, strides,
+                                                                           padding,
+                                                                           out_dtype)})
+    return output
 
-# register original implementation of depthwise_conv2d_nchw since we don't need to change this part
-autotvm.register_topi_compute(depthwise_conv2d_nchw, ['arm_cpu', 'cpu'], 'direct',
-                              depthwise_conv2d_nchw.fdefault)
 
 # register customized schedule for arm cpu.
-@autotvm.register_topi_schedule(schedule_depthwise_conv2d_nchw, ['arm_cpu', 'cpu'], 'direct')
+@autotvm.task.register_topi_schedule(schedule_depthwise_conv2d_nchw, 'arm_cpu', 'direct')
```

Comment:
@FrozenGene Would you mind also test how this change impact Intel CPU, as I described in #2045 ?

Reply:
I will test it on my Mac cpu. What you want to compare? Previous ARM CPU Depthwise convolution schedule on x86 or your PR #2045? In fact, I tested NCHW[x]c schedule on ARM before, which doesn't have performance advantages on Mobilenet V1 model and even slower than previous depthwise convolution schedule.

Reply:
I would like to clarify. I'm not saying we should use NCHW[x]c on ARM. What I wondering is,
1. You remove the register for "cpu" here, it will make x86 use the default schedule for `depthwise_conv2d_nchw`, and users will see performance decreases. So we need to either add the register back, or make another for x86.
2. Then the question is, shall we rely on this schedule for x86, or make another nchw one? I see great performance decreasing on x86 with this schedule, compared to the previous one, both are using fallback config - so I suggest that, if you could provide the tuned fallback schedule, we can test it on x86 to see what is the better path forward.

Reply:
Ok, I understand your concern. But my concern is that previous depthwise convolution fallback will load reference log: 
```python
        if cfg.is_fallback:
            ref_log = autotvm.tophub.load_reference_log(
                'arm_cpu', 'rk3399', 'depthwise_conv2d_nchw', 'direct')
            cfg.fallback_with_reference_log(ref_log)
```
However, we haven't uploaded such tuned log for this PR's schedule(I think this is the main reason of performance decreasing on x86). 

I can not make sure that whether I upload the tuned log for this PR's schedule before merge it for your fallback testing. But from my side, I can test that: tune for arm_cpu  using my schedule on mobilenet v1 and store it on one log file, then x86 use this log file to be fallback.

Reply:
We can not compare templates with fallback configs. So we can tell nothing from @yzlius 's test on x86.

The old template have good fallback mechanism, while the new template does not have any fallback mechanism. So in this case, the new template just gives naive implementation with no tiling, vectorization, etc.

Reply:
@merrymercy yes, I understand your meaning. @yzhliu have added `cpu` on this new schedule template but have poorer performance on x86 with fallback config. However, to verify it, I want to do is to add one tuned log for arm_cpu using new template,  then make it be the standard of x86 fallback like old template has.

Reply:
> It is not easy to do so automatically. You can just paste your log in a comment. But It also does not make much sense to test arm_cpu log on x86.
> So the only reasonable way for @yzhliu to verify is to tune by himself.

I know it. However, previous schedule does like this. Maybe we should have one `x86` cpu fallback log to reference, not `arm_cpu` log to reference like old schedule.

Reply:
@yzhliu could you tune it use this schedule? I quickly test it on my x86 cpu, the first depthwise convolution gflops is 23.75GFLOPS, while previous schedule is 14.44GFLOPS. I think I would agree @merrymercy’s point,  it does not make sense use arm_cpu log to be fallback test on x86 cpu.

Reply:
Sure, I'll tune it today and report the number. Thanks for clarifying, I got a more clear picture now.

Reply:
With tuned kernel on x86, I got 5.53ms (O2) and 5.68ms (O3) on c5.9xlarge for mobilenet, while the previous one is 2.2ms (O2) and 2.7ms (O3). (It's normal to have O2 better than O3 in this case since O3 will convert non-depthwise conv2d to NCHW[x]c, thus introduces layout transformer between depthwise-conv and non-depthwise-conv)

Here's my script, config and running log:

https://gist.github.com/yzhliu/fe28f0e9add3cbd70dda4c7cd30c1275
https://gist.github.com/yzhliu/c287798592f7568c6e624b33727576a5
https://gist.github.com/yzhliu/dd7e16fbe3599ba5dce14986eb699d60

Let me know if I did something wrong. The LLVM errors in the running log looks a little bit weird.

I'll tune again tonight, coz I just realized I made a mistake that I tuned using only 1-thread (but evaluated with 18). Not sure whether that would make difference on final performance.

Reply:
@FrozenGene I'll try your option. How did you calculate gflops for mobilenet? it can be implied from running time, right?

Reply:
@yzhliu I mean the tunning log's GFLOPS. For example, I list the GFLOPS here:
```
Currently:
[Task 2/20] Current/Best: 0.98/ 2.32 GFLOPS | Progress: (1427/2000) | 2679.82 s Done.

After this PR optimization:
[Task 2/20] Current/Best: 0.00/ 4.83 GFLOPS | Progress: (1682/2000) | 1470.40 s Done
```
Then I can compare the best GFLOPS between old / new schedule. And the running time should also affect this.

For the LLVM error, I also met this. You can redirect log python3 tune.py > gflops.log, without 2>& 1.

Reply:
@yzhliu This gist file is on my x86 machine new schedule v.s. previous schedule on mobilenet v1's depthwise tunning GFLOPS log: https://gist.github.com/FrozenGene/2142bbe9c4df1ca307328a53cdac7b05

At my x86 machine, this new PR's schedule is better than old schedule. You could tune again after setting my option and correct thread nums.

Reply:
got it. thanks for clarifying. will do.

Reply:
Now it's 1.87ms (O2) and 2.09ms (O3), pretty cool!
new log: https://gist.github.com/yzhliu/c5b4ad171c0d8b862fe30bdad09434b3

I think then we can add the "cpu" register back, and upload this log to tophub.

Reply:
Good. it is better than old depthwise schedule on x86 too. However, for the fallback, I think it is also not make much sense using arm_cpu log. So do you have suggestion on it? @merrymercy @yzhliu

Reply:
what do you mean by "arm_cpu" log?

Reply:
> what do you mean by "arm_cpu" log?

I mean this code:
```python
        if cfg.is_fallback:
            ref_log = autotvm.tophub.load_reference_log(
                'arm_cpu', 'rk3399', 'depthwise_conv2d_nchw', 'direct')  # This is arm_cpu log
            cfg.fallback_with_reference_log(ref_log)
```
However, we use this log to be fallback reference log before.

Reply:
I believe we should have new models for fallback log, per cpu and arm_cpu